code/parachain/runtime/common/src/dex.rs

```diff
@@ -0,0 +1,262 @@
+use crate::{cosmwasm::*, prelude::*};
+use composable_traits::dex::*;
+use cosmwasm_std::{to_binary, Coin, QueryResponse, Response, Uint128};
+use primitives::currency::CurrencyId;
+use sp_runtime::traits::Convert;
+use sp_std::marker::PhantomData;
+
+pub struct DexPrecompile<Balance, AccountId, Dex>(PhantomData<(Balance, AccountId, Dex)>);
+
+pub enum CosmwasmPrecompileError {
+	AssetConversion,
+	AccountConversion,
+	DispatchError,
+	Serde,
+}
+
+impl<Dex, Balance, AccountId> DexPrecompile<Balance, AccountId, Dex>
+where
+	Dex:
+		Amm<AssetId = CurrencyId, Balance = Balance, AccountId = AccountId, PoolId = crate::PoolId>,
+	Balance: From<Uint128> + Into<u128>,
+	AccountId: sp_std::convert::From<sp_runtime::AccountId32>,
+{
+	fn to_substrate(
+		in_asset: &Coin,
+	) -> Result<AssetAmount<CurrencyId, Balance>, CosmwasmPrecompileError> {
+		let in_asset_id = CosmwasmToSubstrateAssetId::convert(in_asset.denom.clone())
+			.map_err(|_| CosmwasmPrecompileError::AssetConversion)?;
+		let in_amount: Balance = in_asset.amount.into();
+		Ok(AssetAmount::new(in_asset_id, in_amount))
+	}
+
+	fn to_cw(amount: AssetAmount<CurrencyId, Balance>) -> Coin {
+		Coin {
+			denom: CosmwasmToSubstrateAssetId::convert(amount.asset_id),
+			amount: amount.amount.into().into(),
+		}
+	}
+
+	pub fn query(sender: &str, msg: QueryMsg) -> Result<QueryResponse, CosmwasmPrecompileError> {
+		match msg {
+			QueryMsg::Assets { pool_id } => {
```

Comment:
Could you extract each arm to a separate method, something like:

```
	pub fn query(sender: &str, msg: QueryMsg) -> Result<QueryResponse, CosmwasmPrecompileError> {
		match msg {
			QueryMsg::Assets { pool_id } => self.handle_query_assets(sender, pool_id),
			QueryMsg::SpotPrice { pool_id, base_asset, quote_asset_id, calculate_with_fees } => {
				handle_query_spot_price(sender, pool_id, base_asset, quote_asset_id, calculate_with_fees)
			},
	/* ... */
```

and same with `execute`?  The long functions and deep indentation
levels make the code less readable than it could be.

Reply:
handle_query_spot_price will require generics and will make more code.

and you are wrong, consider `QueryMsg::SpotPrice ` to be inline function. and block => { } to be its body.

Adding function is just noise. So your solution is opinionted, adds more noise, more code, harder to read.

Reply:
Why would it require generics? It would be no more generic than the `query` method.

It adds noise because you have to define separate methods with arguments, but the point is that you end up with much shorter functions with cleaner control flow. The functions can also be properly documented with a comment.

Reply:
control flow is not cleaner. match = functions.

Reply:
functions are documented on query types. query case = function. match block = function.

Reply:
generic would be required on function declaration.

Reply:
each variant in CW enum is functions. So when you match, you do functions. And each functions has name (in match variant) and each block of match if function body. So it is same thing encoded other way.

Reply:
They documented as comments to CW people where needed.